SOLR-16853 Upgrade Lucene to 9.7.0

stillalex opened the pull request:
https://issues.apache.org/jira/browse/SOLR-16853

<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

For something minor (i.e. that wouldn't be worth putting in release notes), you can skip JIRA. 
To create a Jira issue, you will need to create an account there first.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

Upgrade Lucene to 9.7.0

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

stillalex commented:
running the tests again for good measure

stillalex commented:
this is failing but due to unrelated changes 
```
> Task :solr:core:compileTestJava
/home/runner/work/solr/solr/solr/core/src/test/org/apache/solr/handler/admin/api/CoreReplicationAPITest.java:88: warning: [ClassCanBeStatic] Inner class is non-static but does not reference enclosing class
  class CoreReplicationAPIMock extends CoreReplicationAPI {
  ^
    (see https://errorprone.info/bugpattern/ClassCanBeStatic)
  Did you mean 'static class CoreReplicationAPIMock extends CoreReplicationAPI {'?
error: warnings found and -Werror specified

> Task :solr:core:compileTestJava FAILED
```

alessandrobenedetti reviewed (approved):
I was thinking it was much more work, it actually seems super straightforward!
+1 on this, as soon as we get another +1 I'll suggest to merge!

stillalex commented:
thank you @alessandrobenedetti for the review!
build is failing due to unrelated changes. rebased and pushed again. added changes.txt entry.
is there anything specific we need to add to Solr startup script for enabling vector support?

stillalex commented:
reading the lucene change log, the following needs to be added to Solr startup

```
Applications started with command line parameter "java --add-modules jdk.incubator.vector" on exactly Java 20 or 21 will automatically use the new vectorized implementations if running on a supported platform
```

HoustonPutman commented:
Yes, but let's keep that as a separate Ticket & PR.

This should be good to go.

HoustonPutman commented:
I've created https://issues.apache.org/jira/browse/SOLR-16869 to track the vector optimizations
